google/cloud/aiplatform/base.py

```diff
@@ -16,13 +16,196 @@
+
+        Args:
+            method (Callable): Required. The method to submit.
+            args (Sequence): Required. The arguments to call the method with.
+            kwargs (dict): Required. The keyword arguments to call the method with.
+            additional_dependencies (Optional[Sequence[futures.Future]]):
+                Optional. Additional dependent futures to wait on before executing
+                method. Note: No validation is done on the dependencies.
+            callbacks (Optional[Sequence[Callable[[futures.Future], Any]]]):
+                Optional. Additional Future callbacks to execute once this created
+                Future is complete.
+
+        Returns:
+            future (Future): Future of the submitted method call.
+        """
+
+        def wait_for_dependencies_and_invoke(
+            deps: Optional[Sequence[futures.Future]],
+            method: Callable[..., Any],
+            args: Sequence[Any],
+            kwargs: Dict[str, Any],
+            internal_callbacks: Callable[[Any], Any],
+        ) -> Any:
+            """Wrapper method to wait on any dependencies before submitting method.
+
+            Args:
+                deps (Sequence[futures.Future]):
+                    Required. Dependent futures to wait on before executing method.
+                    Note: No validation is done on the dependencies.
+                method (Callable): Required. The method to submit.
+                args (Sequence[Any]): Required. The arguments to call the method with.
+                kwargs (Dict[str, Any]):
+                    Required. The keyword arguments to call the method with.
+                internal_callbacks: (Callable[[Any], Any]):
+                    Callbacks that take the result of method.
+
+            """
+            # wait for all dependencies to complete
+            futures_results = futures.wait(deps, return_when=futures.FIRST_EXCEPTION)
+
+            # check for raised exceptions before moving forward
+            for future in futures_results.done:
+                future.result()
+
+            result = method(*args, **kwargs)
+
+            # call callbacks from within future
+            if internal_callbacks:
+                for callback in internal_callbacks:
+                    callback(result)
+
+            return result
+
+        # Retrieves any dependencies from arguments.
+        deps = [
+            arg._latest_future
+            for arg in list(args) + list(kwargs.values())
+            if isinstance(arg, FutureManager) and arg._latest_future
+        ]
+
+        if additional_dependencies:
+            deps.extend(additional_dependencies)
+
+        with self.__latest_future_lock:
+
+            # form a dependency on the latest future of this object
+            if self.__latest_future:
+                deps.append(self.__latest_future)
+
+            self.__latest_future = initializer.global_pool.submit(
```

Comment:
Do we maintain a single global pool that contains all the futures for different resources?  (as opposed to every resource noun instance maintains its own graph of futures)

Does this make debugging more difficult?

Reply:
We maintain a global pool. I'm not sure if it would make debugging less difficult if we maintained a pool per resource because a resource noun's future can be dependent on another resource noun's future. I do think there's additional work that can be done to provide more information about the current state of all futures.

Reply:
Yes that makes sense - it would be great if we have a way to allow the user to inspect all futures of the current session, but that may be more of a feature request.